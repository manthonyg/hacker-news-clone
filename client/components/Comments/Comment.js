/* eslint-disable react/no-danger */
import React, { useState } from "react";
import moment from "moment";
import Heading from "../common/Heading/Heading";
import styled from "styled-components";
import { Link } from "react-router-dom";
import { truncateString } from "../../utils/truncate";
import { fadeInLeft, fadeInRight } from "../../utils/keyframes/fadeAnimations";
import Button from "../common/Button/Button";
import PropTypes from "prop-types";

const CommentCard = styled.div`
  opacity: 0;
  width: 100%;
  min-height: 100px;
  animation: ${fadeInLeft};
  animation-duration: 0.3s;
  animation-delay: ${(props) => props.animationOrder}ms;
  animation-timing-function: cubic-bezier(0.39, 0.575, 0.565, 1);
  animation-fill-mode: both;
  background-color: ${(props) => {
    if (props.theme.theme === "light") return "#FFFFFF";
    return "#282828";
  }};
  border-bottom: ${(props) => {
    if (props.theme.theme === "light") return "1px solid #c3c3c3";
    return "1px solid #404040";
  }};
  &:nth-child(even) {
    animation: ${fadeInRight};
    animation-duration: 0.3s;
    animation-delay: ${(props) => props.animationOrder}ms;
    animation-timing-function: cubic-bezier(0.39, 0.575, 0.565, 1);
    animation-fill-mode: both;
  }
`;

const StyledLink = styled(Link)`
  text-decoration: underline;
  color: ${(props) => {
    if (props.theme.theme === "light") return "#404040";
    return "#bb86fc";
  }};
`;

const CommentCardHeader = styled.p`
  margin-left: 5px;
  margin-top: 10px;
`;

const CommentCardContent = styled.p`
  display: block;
  flex-wrap: wrap;
  overflow-wrap: break-word;
  max-width: 800px;
  margin: 10px 10%;
  position: relative;
  height: 100%;
  &: after {
  content: "";
  text-align: right;
  position: absolute;
  bottom: 0;
  right: 0;
  width: 100%;
  height: 100%;
  background: ${(props) => {
    if (!!props.truncatedStyle) {
      if (props.theme.theme === "light") {
        return "linear-gradient(to bottom, rgba(255, 255, 255, 0), rgba(255, 255, 255, 1) 100%)";
      }
      return "linear-gradient(to bottom, rgba(40, 40, 40, 0), rgba(40, 40, 40, 1) 100%)";
    }
  }}
`;

function Comment({ comment }) {
  const { by, id, text, time } = comment;

  const MAX_STRING_LENGTH = 500;
  const [isTruncated, setIsTruncated] = useState(
    text?.length > MAX_STRING_LENGTH
  );

  const handleTruncate = () => {
    setIsTruncated(!isTruncated);
  };

  return (
    <CommentCard>
      <CommentCardHeader>
        <Heading h5>
          <StyledLink to={`/user?id=${by}`}>{by}</StyledLink> at{" "}
          {moment.unix(time).format("YYYY-MM-DD")}:
        </Heading>
      </CommentCardHeader>{" "}
      <CommentCardContent truncatedStyle={isTruncated}>
        {" "}
        <p
          dangerouslySetInnerHTML={{
            __html:
              text?.length > MAX_STRING_LENGTH
                ? truncateString(text, MAX_STRING_LENGTH, isTruncated)
                : text,
          }}
        />
      </CommentCardContent>
      {text?.length > MAX_STRING_LENGTH && (
        <Button onClick={handleTruncate}>
          {isTruncated ? "read more" : "less"}
        </Button>
      )}
    </CommentCard>
  );
}

Comment.propTypes = {
  comment: PropTypes.shape({
    id: PropTypes.number.isRequired,
    by: PropTypes.string.isRequired,
    text: PropTypes.string.isRequired,
    time: PropTypes.number.isRequired,
  }),
};

export default Comment;
