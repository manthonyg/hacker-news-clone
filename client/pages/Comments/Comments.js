/* eslint-disable no-console */
import React, { useState, useEffect, useRef } from "react";
import queryString from "query-string";
import Loader from "../../components/common/Loader/Loader";
import { fetchComments, fetchItem } from "../../utils/api";
import Flex from "../../components/common/Flex/Flex";
import Heading from "../../components/common/Heading/Heading";
import PostSkeleton from "../Posts/PostSkeleton";
import Post from "../../components/Post/Post";
import CommentList from "../../components/CommentList/CommentList";
import useInfiniteScroll from "../../hooks/useInfiniteScroll";
import ScrollArrow from "../../components/ScrollArrow/ScrollArrow";

function Comments() {
  const INFINITE_SCROLL_FETCH_AMOUNT = 10;
  const { id } = queryString.parse(location.search);
  const [comments, setComments] = useState([]);
  const [commentIds, setCommentIds] = useState([]);
  const [post, setPost] = useState(null);
  const [isLoading, setIsLoading] = useState(false);
  const [isFetching, setIsFetching] = useInfiniteScroll(() =>
    handleInfiniteScroll()
  );
  const [hasComments, setHasComments] = useState(true);
  const componentIsMounted = useRef(true);

  useEffect(() => {
    if (componentIsMounted.current) {
      setIsLoading(true);
      fetchItem(id)
        .then((post) => {
          setPost(post);
          setCommentIds(post?.kids);
          return fetchComments(
            post?.kids.slice(0, INFINITE_SCROLL_FETCH_AMOUNT) || []
          );
        })
        .then((comments) => {
          setComments(comments);
          setIsLoading(false);
        })
        .catch((error) => {
          console.warn("There was a problem gathering these resources", error);
        });
    }
    return () => {
      componentIsMounted.current = false;
    };
  }, []);

  const checkIfHasComments = () => {
    if (comments.length >= commentIds.length) {
      setHasComments(false);
    }
  };
  const handleInfiniteScroll = () => {
    if (hasComments) {
      setTimeout(() => {
        const currentCommentLength = comments.length;
        fetchComments(
          commentIds.slice(
            currentCommentLength - 1,
            currentCommentLength + INFINITE_SCROLL_FETCH_AMOUNT
          )
        ).then((response) => {
          setComments(comments.concat(response));
          setIsFetching(false);
          checkIfHasComments();
        });
      }, 1000);
    }
  };

  return (
    <>
      {isLoading ? (
        <Flex>
          <Heading h5>viewing:</Heading>
          <PostSkeleton numberOfSkeletons={1} />
          <Heading h5>comments: (...)</Heading>
          <PostSkeleton numberOfSkeletons={10} noEmoji />
        </Flex>
      ) : (
        <>
          <Heading h5>viewing:</Heading>
          {post && <Post post={post} />}
          {comments && (
            <Heading h5 isSticky>
              comments ({comments?.length - 1} of {commentIds?.length}):
            </Heading>
          )}

          <div style={{ overflow: "auto" }}>
            <CommentList comments={comments} />
            {isFetching && hasComments && (
              <>
                <Heading h5>Loading comments...</Heading>
                <Loader />
              </>
            )}

            {!hasComments && <Heading h4>no more comments</Heading>}
          </div>
          <ScrollArrow />
        </>
      )}
    </>
  );
}

export default Comments;
