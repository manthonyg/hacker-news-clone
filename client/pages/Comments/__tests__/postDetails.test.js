import React from "react";
import renderer from "react-test-renderer";
import Comments from "../Comments";

it("renders correctly with all props", () => {
  const requiredProps = {};
  const tree = renderer
    .create(
      <>
        <Comments {...requiredProps} />
      </>
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});
