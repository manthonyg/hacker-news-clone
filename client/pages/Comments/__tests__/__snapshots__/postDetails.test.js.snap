// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly with all props 1`] = `
Array [
  <h5
    className="sc-fzozJi hcnTuc"
  >
    viewing:
  </h5>,
  <h5
    className="sc-fzozJi kDKfcF"
  >
    comments (
    -1
     of 
    0
    ):
  </h5>,
  <div
    style={
      Object {
        "overflow": "auto",
      }
    }
  />,
  <svg
    className="sc-fzqARJ iExvbz"
    fill="currentColor"
    height="1em"
    onClick={[Function]}
    stroke="currentColor"
    strokeWidth="0"
    style={
      Object {
        "color": undefined,
        "display": "none",
        "height": 40,
      }
    }
    viewBox="0 0 512 512"
    width="1em"
    xmlns="http://www.w3.org/2000/svg"
  >
    <path
      d="M8 256C8 119 119 8 256 8s248 111 248 248-111 248-248 248S8 393 8 256zm143.6 28.9l72.4-75.5V392c0 13.3 10.7 24 24 24h16c13.3 0 24-10.7 24-24V209.4l72.4 75.5c9.3 9.7 24.8 9.9 34.3.4l10.9-11c9.4-9.4 9.4-24.6 0-33.9L273 107.7c-9.4-9.4-24.6-9.4-33.9 0L106.3 240.4c-9.4 9.4-9.4 24.6 0 33.9l10.9 11c9.6 9.5 25.1 9.3 34.4-.4z"
    />
  </svg>,
]
`;
